File: watch-apps/wear-os/app/src/main/java/com/taishanglaojun/wearos/services/LocationService.java
package com.taishanglaojun.wearos.services;

import android.app.Service;
import android.content.Intent;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.IBinder;
import android.util.Log;
import android.content.Context;
import android.os.Bundle;

/**
 * 位置追踪服务
 * 负责获取用户位置信息并上传到服务器
 */
public class LocationService extends Service implements LocationListener {
    
    private static final String TAG = "LocationService";
    private static final long MIN_TIME_BETWEEN_UPDATES = 10000; // 10秒
    private static final float MIN_DISTANCE_CHANGE_FOR_UPDATES = 10; // 10米
    
    private LocationManager locationManager;
    private boolean isLocationServiceRunning = false;
    
    @Override
    public void onCreate() {
        super.onCreate();
        Log.d(TAG, "位置服务创建");
        
        locationManager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
    }
    
    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.d(TAG, "位置服务启动");
        
        startLocationTracking();
        
        return START_STICKY; // 服务被杀死后自动重启
    }
    
    /**
     * 开始位置追踪
     */
    private void startLocationTracking() {
        if (isLocationServiceRunning) {
            Log.d(TAG, "位置服务已在运行");
            return;
        }
        
        try {
            // 检查权限
            if (checkSelfPermission(android.Manifest.permission.ACCESS_FINE_LOCATION) 
                != android.content.pm.PackageManager.PERMISSION_GRANTED) {
                Log.e(TAG, "缺少位置权限");
                return;
            }
            
            // 请求位置更新
            if (locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
                locationManager.requestLocationUpdates(
                    LocationManager.GPS_PROVIDER,
                    MIN_TIME_BETWEEN_UPDATES,
                    MIN_DISTANCE_CHANGE_FOR_UPDATES,
                    this
                );
                isLocationServiceRunning = true;
                Log.d(TAG, "GPS位置追踪已启动");
            }
            
            if (locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER)) {
                locationManager.requestLocationUpdates(
                    LocationManager.NETWORK_PROVIDER,
                    MIN_TIME_BETWEEN_UPDATES,
                    MIN_DISTANCE_CHANGE_FOR_UPDATES,
                    this
                );
                Log.d(TAG, "网络位置追踪已启动");
            }
            
        } catch (SecurityException e) {
            Log.e(TAG, "位置权限异常: " + e.getMessage());
        }
    }
    
    /**
     * 停止位置追踪
     */
    private void stopLocationTracking() {
        if (locationManager != null && isLocationServiceRunning) {
            try {
                locationManager.removeUpdates(this);
                isLocationServiceRunning = false;
                Log.d(TAG, "位置追踪已停止");
            } catch (SecurityException e) {
                Log.e(TAG, "停止位置追踪异常: " + e.getMessage());
            }
        }
    }
    
    @Override
    public void onLocationChanged(Location location) {
        Log.d(TAG, "位置更新: " + location.getLatitude() + ", " + location.getLongitude());
        
        // 处理位置数据
        processLocationData(location);
    }
    
    /**
     * 处理位置数据
     */
    private void processLocationData(Location location) {
        // 创建位置数据对象
        LocationData locationData = new LocationData(
            location.getLatitude(),
            location.getLongitude(),
            location.getAltitude(),
            location.getAccuracy(),
            System.currentTimeMillis()
        );
        
        // 上传到服务器
        uploadLocationData(locationData);
    }
    
    /**
     * 上传位置数据到服务器
     */
    private void uploadLocationData(LocationData locationData) {
        // TODO: 实现数据上传逻辑
        Log.d(TAG, "位置数据准备上传: " + locationData.toString());
    }
    
    @Override
    public void onStatusChanged(String provider, int status, Bundle extras) {
        Log.d(TAG, "位置提供者状态变化: " + provider + ", 状态: " + status);
    }
    
    @Override
    public void onProviderEnabled(String provider) {
        Log.d(TAG, "位置提供者启用: " + provider);
    }
    
    @Override
    public void onProviderDisabled(String provider) {
        Log.d(TAG, "位置提供者禁用: " + provider);
    }
    
    @Override
    public void onDestroy() {
        super.onDestroy();
        stopLocationTracking();
        Log.d(TAG, "位置服务销毁");
    }
    
    @Override
    public IBinder onBind(Intent intent) {
        return null; // 不支持绑定
    }
    
    /**
     * 位置数据内部类
     */
    private static class LocationData {
        public final double latitude;
        public final double longitude;
        public final double altitude;
        public final float accuracy;
        public final long timestamp;
        
        public LocationData(double lat, double lng, double alt, float acc, long time) {
            this.latitude = lat;
            this.longitude = lng;
            this.altitude = alt;
            this.accuracy = acc;
            this.timestamp = time;
        }
        
        @Override
        public String toString() {
            return String.format("LocationData{lat=%.6f, lng=%.6f, alt=%.2f, acc=%.2f, time=%d}",
                latitude, longitude, altitude, accuracy, timestamp);
        }
    }
}

File: watch-apps/wear-os/app/src/main/java/com/taishanglaojun/watch/service/LocationService.kt
package com.taishanglaojun.watch.service

import android.Manifest
import android.app.Service
import android.content.Intent
import android.content.pm.PackageManager
import android.location.Location
import android.os.IBinder
import androidx.core.app.ActivityCompat
import com.google.android.gms.location.*
import com.google.android.gms.tasks.Task
import com.taishanglaojun.watch.data.repository.TaskRepository
import com.taishanglaojun.watch.data.models.TaskCoordinate
import com.taishanglaojun.watch.data.models.WatchTask
import com.taishanglaojun.watch.services.ConnectivityService
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import javax.inject.Inject
import kotlin.math.*

@AndroidEntryPoint
class LocationService : Service() {

    @Inject
    lateinit var taskRepository: TaskRepository

    @Inject
    lateinit var connectivityService: ConnectivityService

    private lateinit var fusedLocationClient: FusedLocationProviderClient
    private lateinit var geofencingClient: GeofencingClient
    
    private val serviceScope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    
    // Location state
    private val _currentLocation = MutableStateFlow<Location?>(null)
    val currentLocation: StateFlow<Location?> = _currentLocation.asStateFlow()
    
    private val _isTracking = MutableStateFlow(false)
    val isTracking: StateFlow<Boolean> = _isTracking.asStateFlow()
    
    private val _nearbyTasks = MutableStateFlow<List<WatchTask>>(emptyList())
    val nearbyTasks: StateFlow<List<WatchTask>> = _nearbyTasks.asStateFlow()
    
    // Location request configuration
    private val locationRequest = LocationRequest.Builder(
        Priority.PRIORITY_HIGH_ACCURACY,
        10000L // 10 seconds
    ).apply {
        setMinUpdateIntervalMillis(5000L) // 5 seconds
        setMaxUpdateDelayMillis(30000L) // 30 seconds
    }.build()
    
    private val locationCallback = object : LocationCallback() {
        override fun onLocationResult(locationResult: LocationResult) {
            locationResult.lastLocation?.let { location ->
                _currentLocation.value = location
                checkNearbyTasks(location)
                
                // Send location update to phone if connected
                serviceScope.launch {
                    if (connectivityService.connectionState.value.isConnected) {
                        // Send location update (implement as needed)
                    }
                }
            }
        }
    }
    
    override fun onCreate() {
        super.onCreate()
        fusedLocationClient = LocationServices.getFusedLocationProviderClient(this)
        geofencingClient = LocationServices.getGeofencingClient(this)
    }
    
    override fun onBind(intent: Intent?): IBinder? = null
    
    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        when (intent?.action) {
            ACTION_START_TRACKING -> startLocationTracking()
            ACTION_STOP_TRACKING -> stopLocationTracking()
            ACTION_UPDATE_GEOFENCES -> updateGeofences()
        }
        return START_STICKY
    }
    
    override fun onDestroy() {
        super.onDestroy()
        stopLocationTracking()
        serviceScope.cancel()
    }
    
    private fun startLocationTracking() {
        if (!hasLocationPermission()) {
            return
        }
        
        try {
            fusedLocationClient.requestLocationUpdates(
                locationRequest,
                locationCallback,
                null
            )
            _isTracking.value = true
            
            // Get last known location
            fusedLocationClient.lastLocation.addOnSuccessListener { location ->
                location?.let {
                    _currentLocation.value = it
                    checkNearbyTasks(it)
                }
            }
            
        } catch (e: SecurityException) {
            // Handle permission error
        }
    }
    
    private fun stopLocationTracking() {
        fusedLocationClient.removeLocationUpdates(locationCallback)
        _isTracking.value = false
    }
    
    private fun hasLocationPermission(): Boolean {
        return ActivityCompat.checkSelfPermission(
            this,
            Manifest.permission.ACCESS_FINE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED
    }
    
    private fun checkNearbyTasks(location: Location) {
        serviceScope.launch {
            try {
                val allTasks = taskRepository.getAllTasks().value
                val nearby = allTasks.filter { task ->
                    task.coordinate?.let { coord ->
                        val distance = calculateDistance(
                            location.latitude,
                            location.longitude,
                            coord.latitude,
                            coord.longitude
                        )
                        distance <= PROXIMITY_THRESHOLD_METERS
                    } ?: false
                }
                
                _nearbyTasks.value = nearby
                
                // Check for task arrival notifications
                nearby.forEach { task ->
                    if (shouldNotifyTaskArrival(task, location)) {
                        notifyTaskArrival(task)
                    }
                }
                
            } catch (e: Exception) {
                // Handle error
            }
        }
    }
    
    private fun calculateDistance(
        lat1: Double, lon1: Double,
        lat2: Double, lon2: Double
    ): Double {
        val earthRadius = 6371000.0 // Earth radius in meters
        
        val dLat = Math.toRadians(lat2 - lat1)
        val dLon = Math.toRadians(lon2 - lon1)
        
        val a = sin(dLat / 2).pow(2) +
                cos(Math.toRadians(lat1)) * cos(Math.toRadians(lat2)) *
                sin(dLon / 2).pow(2)
        
        val c = 2 * atan2(sqrt(a), sqrt(1 - a))
        
        return earthRadius * c
    }
    
    private fun shouldNotifyTaskArrival(task: WatchTask, location: Location): Boolean {
        // Check if we should notify about arriving at task location
        // This could be based on task status, user preferences, etc.
        return task.status == com.taishanglaojun.watch.model.TaskStatus.ACCEPTED ||
               task.status == com.taishanglaojun.watch.model.TaskStatus.IN_PROGRESS
    }
    
    private fun notifyTaskArrival(task: WatchTask) {
        // Send notification about arriving at task location
        // This would integrate with the notification system
    }
    
    private fun updateGeofences() {
        if (!hasLocationPermission()) {
            return
        }
        
        serviceScope.launch {
            try {
                val tasks = taskRepository.getAllTasks().value
                val geofenceRequests = tasks.mapNotNull { task ->
                    task.coordinate?.let { coord ->
                        createGeofenceRequest(task.id, coord)
                    }
                }
                
                if (geofenceRequests.isNotEmpty()) {
                    // Remove existing geofences first
                    geofencingClient.removeGeofences(geofenceRequests.map { it.requestId })
                    
                    // Add new geofences
                    val geofencingRequest = GeofencingRequest.Builder().apply {
                        setInitialTrigger(GeofencingRequest.INITIAL_TRIGGER_ENTER)
                        addGeofences(geofenceRequests)
                    }.build()
                    
                    // Note: This would need a PendingIntent for geofence transitions
                    // geofencingClient.addGeofences(geofencingRequest, geofencePendingIntent)
                }
                
            } catch (e: Exception) {
                // Handle geofencing error
            }
        }
    }
    
    private fun createGeofenceRequest(taskId: String, coordinate: TaskCoordinate): Geofence {
        return Geofence.Builder()
            .setRequestId(taskId)
            .setCircularRegion(
                coordinate.latitude,
                coordinate.longitude,
                GEOFENCE_RADIUS_METERS
            )
            .setExpirationDuration(Geofence.NEVER_EXPIRE)
            .setTransitionTypes(
                Geofence.GEOFENCE_TRANSITION_ENTER or 
                Geofence.GEOFENCE_TRANSITION_EXIT
            )
            .build()
    }
    
    // Public API methods
    fun getCurrentLocation(): Location? = _currentLocation.value
    
    fun getNearbyTasks(): List<WatchTask> = _nearbyTasks.value
    
    fun getDistanceToTask(task: WatchTask): Double? {
        val currentLoc = _currentLocation.value
        val taskCoord = task.coordinate
        
        return if (currentLoc != null && taskCoord != null) {
            calculateDistance(
                currentLoc.latitude,
                currentLoc.longitude,
                taskCoord.latitude,
                taskCoord.longitude
            )
        } else null
    }
    
    companion object {
        const val ACTION_START_TRACKING = "start_tracking"
        const val ACTION_STOP_TRACKING = "stop_tracking"
        const val ACTION_UPDATE_GEOFENCES = "update_geofences"
        
        private const val PROXIMITY_THRESHOLD_METERS = 100.0
        private const val GEOFENCE_RADIUS_METERS = 50.0f
    }
}